### ISSUE TITLE
Not enough building merging & simplification

### ISSUE THREAD

**Issue by zerebubuth** [tilezen/vector-datasource#1686]
Building heights and extra properties (e.g: `roof_shape` or `roof_colour`) are preventing merging between buildings. For example, tile `14/6066/9294` in the north-west of Sao Paulo:

![North-west Sao Paulo, coloured by unique feature](https://user-images.githubusercontent.com/271360/46810583-eb228100-cd68-11e8-99f7-3a87bc23d425.png)

(Coloured uniquely by feature.) There is some merging going, but still leaving us with over 9,000 unique multipolygons. Partly, this is because we're not quantising height enough:

![North-west Sao Paulo, coloured by height](https://user-images.githubusercontent.com/271360/46810828-671cc900-cd69-11e8-9d32-c650c1785de7.png)

(There are 5,406 unique height values.) Even when we zoom out, and are quantising values (although not all?), there's still a lot of merging inhibited by neighbouring building heights differing by more than 5m.

![North-west Sao Paulo, zoom 13, coloured by unique height](https://user-images.githubusercontent.com/271360/46811537-e52d9f80-cd6a-11e8-980c-2d38824f4da7.png)

It also looks like we're not quantising heights for buildings (should they be `building:part`s?) with `min_height` or `layer != 0`.

Also, we have quite a lot of small courtyards within buildings which probably aren't visible at this zoom level.

Finally, in this zoomed-in block from the `13/3033/4647` tile, you can see there's not a lot of merging going on between building geometries (even if they might be merged into the same multipolygon feature), so we're wasting a lot of bytes to describe each individual polygon.

![Overzoomed section of zoom 13 tile, coloured by unique feature](https://user-images.githubusercontent.com/271360/46812198-1a86bd00-cd6c-11e8-8b0f-52601ec08c73.png)

In summary:

- [ ] Drop superfluous properties, e.g: `roof_material`, and quantize height more aggressively at lower zooms for smaller buildings.
- [ ] Make sure we're quantising heights for buildings with `layer` or `min_height`.
- [ ] Drop small courtyards within buildings.
- [ ] Dilate or buffer building geometry slightly while merging to try and merge adjacent buildings with very small gaps between them.

**Comment by nvkelso** [comment]
Looks like the Mapzen house styles hide many buildings at zooms 13 and 14, indicating to me we can drop `scale_rank` 3,4,5 at zoom 13 and `scale_rank` 4,5 at zoom 14.

```
        # building footprints, pre-extrusion
        footprints:
            filter:
                # show footprints for buildings at least one zoom level before they will be extruded
                - { $zoom: 13, scale_rank: [1,2] }
                - { $zoom: 14, scale_rank: [1,2,3] }
                - { $zoom: 15, height: { min: 100 } }
                - { $zoom: 15, area: { min: 500 } }
                - { $zoom: 15, volume: { min: 100000 } }
                - { $zoom: 16, area: { min: 100 } }
                - { $zoom: 16, volume: { min: 50000 } }
                - { $zoom: { min: 17 }, area: true }
```

**Comment by nvkelso** [comment]
Connects to https://github.com/tilezen/vector-datasource/pull/1689.

**Comment by nvkelso** [comment]
A few comments:

First: this change looks visually great for Bubble Wrap and hugely reduces file size. Huzzah!

But there seem to be 2 issues:

- **scale_rank is always 1 now instead of range of 1 to 5.** -- regression we should fix
- **zoom 14 missing some buildings** -- not a ship block issue, but we'd likely fix it fixing the regression, and it's important for a style like **Refill** which we've compromised a bit.

_To clarify, I think zoom 13 looks **better** in this new setup (dropping more buildings), but zoom 14 looks worse (drops slightly too many buildings)._

Let's dig into this...

This [change](https://github.com/tilezen/vector-datasource/pull/1689/files) effectively almost always renders features with `scale_rank` of 1 for merged features (versus pre-merged features), when there used to be range of 0 to 5 – see all the **dark green** in the GIFs above.

> `# assign `scale_rank` - AFTER merging, as this changes the area and volume`
> fn: vectordatasource.transform.csv_match_properties

That's apparently why the **"blue" buildings** in the *before* at zoom 14 are missing from the **after** in, because merged buildings always have huge area: 

SF: 14/37.7822/-122.4300

![buildings_zoom_14](https://user-images.githubusercontent.com/853051/49985803-29335100-ff22-11e8-95fb-df63679ddeab.gif)

Which also makes this "always" true at zooms 13 and 14, because all features are always incorrectly scale_rank: 1:

>                - { $zoom: 13, scale_rank: [1,2] }
>                - { $zoom: 14, scale_rank: [1,2,3] }

Seems like merging needs to happen after `drop_features_where`?

>   - fn: vectordatasource.transform.drop_features_where

And to keep the fewer buildings at zoom 13 we'd change this line to be > 1?

>       where: scale_rank > 2

And the updated **quantize** needs to probably be done FIRST, but MERGING split off from that and done last?

>       13: vectordatasource.transform.quantize_height_round_nearest_20_meters
>        14: vectordatasource.transform.quantize_height_round_nearest_10_meters
>        15: vectordatasource.transform.quantize_height_round_nearest_10_meters

I suspect the new **height** filter (below) makes it mostly look correct, except for zoom 14 and the excessive use of scale_rank: 1?

Like if the old scale_rank [1,2,3] logic (which includes height) for pre-merge building with a height >= 100 then it'll still appear in zoom 14, but in this new setup lower height but larger area/volume buildings are now excluded. 

```
  - fn: vectordatasource.transform.numeric_min_filter
    params:
      source_layer: buildings
      mode: any
      filters:
        13: { area: 5000, height: 150, volume: 150000 }
        14: { area: 3000, height: 100, volume: 100000 }
        15: { area:  500, height: 100, volume: 100000 }
```

Merging gets unhappy when processing too many buildings so we do some partitioning there along the way now... but probably pre-partitioning by `scale_rank` again would accomplish some of those same goals?

(Look at the **"blue" buildings**...)

**Los Angeles** 14/34.0013/-118.4290:

![buildings_zoom_14_los_angeles](https://user-images.githubusercontent.com/853051/49988412-72d46980-ff2b-11e8-82ec-58b067cc98e8.gif)

**Paris** 14/48.8586/2.3264:

![buildings_zoom_14_paris](https://user-images.githubusercontent.com/853051/49988419-7b2ca480-ff2b-11e8-93fc-2454eec2bc30.gif)

**Sao Paulo** 14/-23.5427/-46.6418:

![buildings_zoom_14_sao_paulo](https://user-images.githubusercontent.com/853051/49988422-7f58c200-ff2b-11e8-8624-01e14a05c75f.gif)
